nsxt/resource_nsxt_upgrade_run.go

```diff
@@ -0,0 +1,568 @@
+		if err != nil {
+			return err
+		}
+
+		//Post upgrade check
+		if component != mpUpgradeGroup {
+			settingI := d.Get(componentToSettingKey[component]).([]interface{})
+			if len(settingI) == 0 {
+				continue
+			}
+
+			setting := settingI[0].(map[string]interface{})
+			postCheck := setting["post_upgrade_check"].(bool)
+
+			if postCheck {
+				upgradeClientSet.UpgradeClient.Executepostupgradechecks(component)
+			}
+		}
+	}
+	return nil
+}
+
+func setOutput(upgradeClientSet upgradeClientSet, d *schema.ResourceData) error {
+	results, err := upgradeClientSet.GroupClient.List(nil, nil, nil, nil, nil, nil, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	var plans []map[string]interface{}
+	for _, result := range results.Results {
+		elem := make(map[string]interface{})
+		elem["id"] = result.Id
+		elem["parallel"] = result.Parallel
+		elem["enabled"] = result.Enabled
+		elem["pause_after_each_upgrade_unit"] = result.PauseAfterEachUpgradeUnit
+		elem["type"] = result.Type_
+		if *result.Type_ != mpUpgradeGroup {
+			for _, config := range result.ExtendedConfiguration {
+				for k := range supportedUpgradeGroupExtendedConfigValue {
+					if k == *config.Key {
+						elem[*config.Key] = *config.Value
+					}
+				}
+			}
+		}
+		plans = append(plans, elem)
+	}
+	d.Set("upgrade_plan", plans)
+
+	status, err := upgradeClientSet.StatusClient.Get(nil, nil, nil)
+	if err != nil {
+		return err
+	}
+	var states []map[string]interface{}
+	for _, result := range status.ComponentStatus {
+		elem := make(map[string]interface{}, 5)
+		elem["type"] = result.ComponentType
+		elem["status"] = result.Status
+		elem["percent_complete"] = result.PercentComplete
+		elem["details"] = result.Details
+		elem["target_version"] = result.TargetComponentVersion
+		states = append(states, elem)
+	}
+	d.Set("state", states)
+	return nil
+}
+
+func resourceNsxtUpgradeRunRead(d *schema.ResourceData, m interface{}) error {
+	id := d.Id()
+	connector := getPolicyConnector(m)
+	upgradeClientSet := newUpgradeClientSet(connector)
+	err := setOutput(upgradeClientSet, d)
+	if err != nil {
+		return handleReadError(d, "NsxtUpgradeRun", id, err)
+	}
+	return nil
+}
+
+func resourceNsxtUpgradeRunUpdate(d *schema.ResourceData, m interface{}) error {
+	id := d.Id()
+	connector := getPolicyConnector(m)
+	upgradeClientSet := newUpgradeClientSet(connector)
+
+	log.Printf("[INFO] Updating UpgradeUnitGroup and UpgradePlanSetting.")
+	err := prepareUpgrade(upgradeClientSet, d, false)
+	if err != nil {
+		return handleUpdateError("NsxtUpgradeRun", id, err)
+	}
+
+	log.Printf("[INFO] Successfully update UpgradeUnitGroup and UpgradePlanSetting. Start Upgrade.")
+
+	err = runUpgrade(upgradeClientSet, d)
+	if err != nil {
+		log.Printf("[WARN] Upgrade didn't success: %v. Refer to the state of resource for more information.", err)
```

Comment:
We don't return error here?

Reply:
I didn't return an error here because for example: if edge upgrade timeout, when users "update" the resource we could know if any group settings or plan settings change. If only the host upgrade setting is changed, there is no need to pause-modify-resume the edge upgrade.
The approach of only creating the resource when all upgrades are successful also sounds good. It is more intuitive and easy to maintain.
With the timeout configurable, maybe returning an error here is a better option.

Reply:
Ah my bad. Whether run postcheck should be able to update.

Reply:
I assumed that if, say, an edge was disabled for an upgrade, re-enabling it and applying again should trigger another upgrade cycle? Or do we not support such upgrade paths for now?

Reply:
For your case, only the group that is currently in progress needs to be paused and resume. After re-think, I prefer only creating the resource when all upgrades succeed approach and have committed it.